### PR TITLE
[BEAM-600] Add TestCountingSource option to specify the number of splits returned by generateInitialSplits.

### PR DESCRIPTION
**Pull request by drieber**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by drieber** [comment]
Hi @dhalperi, @dpmills, can you please take a look? Thanks!

**Comment by dpmills** [comment]
LGTM

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java [hunk 1]
```diff
@@ -50,6 +50,8 @@
 
   private static List<Integer> finalizeTracker;
   private final int numMessagesPerShard;
+  // If non-negative, this is the number of splits returned by generateInitialSplits.
+  private final int numSplits;
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java]: no point in having the allowSplitting attribute as well, numSplits subsumes its responsibilities, please remove allowSplitting boolean and just set numSplits to 1 when a person call withoutSplits

Please also add a test to https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSourceTest.java to ensure that withoutSplitting and setting numSplits does what it asks

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java [hunk 2]
```diff
@@ -67,36 +69,42 @@ public static void setFinalizeTracker(List<Integer> finalizeTracker) {
...
+        numMessagesPerShard, shardNumber, dedup, throwOnFirstSnapshot, true, numSplits);
   }
 
   public TestCountingSource withoutSplitting() {
     return new TestCountingSource(
-        numMessagesPerShard, shardNumber, dedup, throwOnFirstSnapshot, false);
+        numMessagesPerShard, shardNumber, dedup, throwOnFirstSnapshot, false, -1);
+  }
+
+  public TestCountingSource withNumSplits(int numSplits) {
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java]: withNumSplits -> withNumberOfInitialSplits

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java (hunk 1 above)
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java]: numSplits -> numInitialSplits

**Comment by lukecwik** [comment]
R: @lukecwik

**Pull request closed without merging**